src/util/retry.rs

```diff
@@ -0,0 +1,226 @@
+/// 		.skip_retry_on_error(|e| matches!(e, VssError::InvalidRequestError(..)));
+///
+/// let result = retry(operation, &retry_policy);
+///```
+pub async fn retry<R, F, Fut, T, E>(mut operation: F, retry_policy: &R) -> Result<T, E>
+where
+	R: RetryPolicy<E = E>,
+	F: FnMut() -> Fut,
+	Fut: Future<Output = Result<T, E>>,
+	E: Error,
+{
+	let mut attempts_made = 0;
+	let mut accumulated_delay = Duration::ZERO;
+	loop {
+		match operation().await {
+			Ok(result) => return Ok(result),
+			Err(err) => {
+				attempts_made += 1;
+				if let Some(delay) =
+					retry_policy.next_delay(&RetryContext { attempts_made, accumulated_delay, error: &err })
+				{
+					tokio::time::sleep(delay).await;
+					accumulated_delay += delay;
+				} else {
+					return Err(err);
+				}
+			}
+		}
+	}
+}
+
+/// Provides the logic for how and when to perform retries.
+pub trait RetryPolicy: Sized {
+	/// The error type returned by the `operation` in `retry`.
+	type E: Error;
+	/// Returns the duration to wait before trying the next attempt.
+	/// `context` represents the context of a retry operation.
+	///
+	/// If `None` is returned then no further retry attempt is made.
+	fn next_delay(&self, context: &RetryContext<Self::E>) -> Option<Duration>;
+
+	/// Returns a new `RetryPolicy` that respects the given maximum attempts.
+	fn with_max_attempts(self, max_attempts: u32) -> MaxAttemptsRetryPolicy<Self> {
+		MaxAttemptsRetryPolicy { inner_policy: self, max_attempts }
+	}
+	/// Returns a new `RetryPolicy` that respects the given total delay.
+	fn with_max_total_delay(self, max_total_delay: Duration) -> MaxTotalDelayRetryPolicy<Self> {
+		MaxTotalDelayRetryPolicy { inner_policy: self, max_total_delay }
+	}
+	/// Returns a new `RetryPolicy` that adds jitter(random delay) to underlying policy.
+	fn with_max_jitter(self, max_jitter: Duration) -> JitteredRetryPolicy<Self> {
+		JitteredRetryPolicy { inner_policy: self, max_jitter }
+	}
+	/// Skips retrying on errors that evaluate to `true` after applying `function`.
+	fn skip_retry_on_error<F>(self, function: F) -> FilteredRetryPolicy<Self, F>
+	where
+		F: 'static + Fn(&Self::E) -> bool,
+	{
+		FilteredRetryPolicy { inner_policy: self, function }
+	}
+}
+
+/// Represents the context of a retry operation.
+///
+/// The context holds key information about the retry operation
+/// such as how many attempts have been made till now, the accumulated
+/// delay between retries, and the error that triggered the retry.
+pub struct RetryContext<'a, E: Error> {
+	/// The number attempts made till now, before attempting the next retry.
+	attempts_made: u32,
+
+	/// The amount of artificial delay we have already waited in between previous
+	/// attempts. Does not include the time taken to execute the operation that is being retried.
+	accumulated_delay: Duration,
+
+	/// The error encountered in the previous attempt.
+	error: &'a E,
+}
+
+/// Decorates the given `RetryPolicy` and provides a `RetryPolicy` that skips retrying errors
```

Comment:
We can assume the concept of decorating is known by the reader. Let's rephrase a little more succinctly as:

```rust
/// Decorates the given `RetryPolicy` by not retrying ...
```

Similar elsewhere.